app/services/contributors_generator.rb

```diff
@@ -0,0 +1,243 @@
+  def self.form_from_contributors(work:)
+    new(work: work).form_from_contributors
+  end
+
+  sig { params(work: Work).void }
+  def initialize(work:)
+    @work = work
+  end
+
+  sig { returns(T::Array[Cocina::Models::Contributor]) }
+  def generate
+    result = []
+
+    work.contributors.each do |work_form_contributor|
+      result << contributor(work_form_contributor)
+    end
+
+    result
+  end
+
+  FORM_EVENT = Cocina::Models::DescriptiveValue.new(
+    {
+      value: 'Event',
+      type: 'resource types',
+      source: { value: 'DataCite resource types' }
+    }
+  )
+
+  sig { returns(T::Array[Cocina::Models::DescriptiveValue]) }
+  def form_from_contributors
+    work.contributors.each do |contributor|
+      next unless contributor.role == 'Conference' || contributor.role == 'Event'
+
+      return [FORM_EVENT]
+    end
+
+    []
+  end
+
+  private
+
+  sig { returns(Work) }
+  attr_reader :work
+
+  sig { params(contributor: Contributor).returns(T.nilable(Cocina::Models::Contributor)) }
+  def contributor(contributor)
+    # FIXME: TODO: mappings for status (primary) and/or order
+    Cocina::Models::Contributor.new(
+      "name": name_descriptive_value(contributor),
+      "type": contributor_type(contributor),
+      "role": cocina_roles(contributor.role)
+    )
+  end
+
+  sig { params(contributor: Contributor).returns(T::Array[Cocina::Models::DescriptiveValue]) }
+  def name_descriptive_value(contributor)
+    return [Cocina::Models::DescriptiveValue.new(value: full_name(contributor))] unless contributor.person?
+
+    [Cocina::Models::DescriptiveValue.new(value: full_name(contributor), type: 'inverted full name')]
+  end
+
+  sig { params(contributor: Contributor).returns(T.nilable(String)) }
+  def full_name(contributor)
+    return "#{contributor.last_name}, #{contributor.first_name}" if contributor.person?
+
+    contributor.full_name
+  end
+
+  sig { params(contributor: Contributor).returns(String) }
+  def contributor_type(contributor)
+    return 'conference' if contributor.role == 'Conference'
+    return 'event' if contributor.role == 'Event'
+
+    contributor.contributor_type
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  sig { params(role: String).returns(T::Array[Cocina::Models::DescriptiveValue]) }
+  def cocina_roles(role)
+    result = [
+      Cocina::Models::DescriptiveValue.new(
+        value: role,
+        source: {
+          value: 'Stanford self-deposit contributor types'
+        }
+      )
+    ]
+    result << T.must(marcrelator_role(role)) if marcrelator_role(role)
+    result << T.must(datacite_role(role)) if datacite_role(role)
+    result
+  end
+  # rubocop:enable Metrics/MethodLength
+
+  # rubocop:disable Metrics/MethodLength
+  sig { params(role: String).returns(T.nilable(Cocina::Models::DescriptiveValue)) }
+  def marcrelator_role(role)
+    mr_code = ROLE_TO_MARC_RELATOR_CODE[role]
+    mr_value = MARC_RELATOR_CODE_TO_VALUE[mr_code]
```

Comment:
I was a little surprised to find these constants down below. I prefer having constants defined together near the top of a class, but I'd also be OK with them located just above the method where they are used since I know that tends to be your preference.